File: php/src/Behavioral/Observer/DvdReleaseByCategory.php
<?

## //DvdReleaseByCategory - the subject
#//(the class which is observed)

class DvdReleaseByCategory
{

	public $categoryName;
	public $subscribers;
	public $releases;

	function __construct($categoryName)
	{
		$this->categoryName = $categoryName;
		$this->subscribers = array();
		$this->releases = array();
	}
	
   
	function addSubscriber($subscriber)
	{
		$this->subscribers[] = $subscriber;
	}
  
	function removeSubscriber($subscriber)
	{
		$i = 0;
		$found_i = -1;
		foreach ( $this->subscribers as $s )
		{
			if($subscriber == $s)
			{
				$found_i = $i;
				break;
			}
			$i++;
		}
		if($found_i != -1 )
		{
			array_splice( $this->subscribers,$found_i,1);
			return 1;
		}
		return 0;
	}
   
	function newDvdRelease($release)
	{
		$this->releases[]= $release;
		return $this->notifySubscribersOfNewDvd($release);
	}
   
	function updateDvd($release)
	{
		$dvdUpdated = 0;
		foreach( $this->releases as $r )
		{	
			if ($release->serialNumber == $r->serialNumber) 
			{
				$r->updateRelease($release); ## add retval of true/false to denote if any data changed -- ie dont notify if no real change was made
				$dvdUpdated = 1;
				break;
			}
		}
		if ($dvdUpdated) 
			return $this->notifySubscribersOfUpdate($release);
		else 
			return $this->newDvdRelease($release);
	}
   
	function notifySubscribersOfNewDvd($release)
	{
		$retval = array();
		foreach( $this->subscribers as $s )
			$retval[] = $s->newDvdRelease($release,$this->categoryName) ;
		return $retval;
	}

	function notifySubscribersOfUpdate($release)
	{
		$retval = array();
		foreach( $this->subscribers as $s)
			$retval[] = $s->updateDvdRelease($release,$this->categoryName) ;
		return $retval;
	}

}

File: go/behavioral/observer/DvdReleaseByCategory.go

//// Original copy of this content taken from http://www.fluffycat.com/Java-Design-Patterns/ in 2010
//// Original Author: Larry Truett
//// Privacy Policy at http://www.fluffycat.com/Privacy-Policy/
//Observer (aka Dependents & Publish-Subscribe) Overview
//An object notifies other object(s) if it changes.
//DvdReleaseByCategory - the subject
//(the class which is observed)


package observer

import (
	"strings"
)


type DvdReleaseByCategory struct {  
   CategoryName string
   subscribers []*DvdSubscriber
   releases []*DvdRelease
}

   
func NewDvdReleaseByCategory(categoryNameIn string) *DvdReleaseByCategory {
	obj := new(DvdReleaseByCategory)
	obj.CategoryName = categoryNameIn
	obj.subscribers = make([]*DvdSubscriber,0)
	obj.releases    = make([]*DvdRelease,0)
	return obj
}

func (this *DvdReleaseByCategory) AddSubscriber( s *DvdSubscriber) {
	 this.subscribers = append(this.subscribers, s)
}

func (this *DvdReleaseByCategory) RemoveSubscriber (s *DvdSubscriber ) bool {
	foundI := -1
	for i,this_s := range this.subscribers {
		if this_s == s {
			foundI = i
			break
		}
	}
	if foundI == -1 {
		return false
	}
	this.subscribers[foundI] = this.subscribers[len(this.subscribers)-1]
	this.subscribers = this.subscribers[:len(this.subscribers)-1]
   return true
}

func (this *DvdReleaseByCategory) NewDvdRelease( r *DvdRelease ) string {
	 this.releases = append(this.releases, r)
	 return this.notifySubscribersOfNewDvd(r)
}

func (this *DvdReleaseByCategory) UpdateDvd( r *DvdRelease ) string {
	dvdUpdated := false
	
	 for _,this_r := range this.releases {
		  if r.SerialNumber == this_r.SerialNumber {
				this_r.UpdateDvdReleaseWithObj( r )
				dvdUpdated = true
				break
		  }
	 }
	 if dvdUpdated {
		  return this.notifySubscribersOfUpdate(r)
	 }
	return this.NewDvdRelease(r)
}

func (this *DvdReleaseByCategory) notifySubscribersOfNewDvd( r *DvdRelease ) string {
	retval := make([]string,0)	
	for _,s := range this.subscribers {
		retval = append(retval, s.NewDvdRelease( r, this.CategoryName ) )
	}
	return strings.Join(retval,"\n")
}

func (this *DvdReleaseByCategory) notifySubscribersOfUpdate( r *DvdRelease ) string {
	retval := make([]string,0)	
	for _,s := range this.subscribers {
		retval = append(retval, s.UpdateDvdRelease( r, this.CategoryName ) )
	}       
	return strings.Join(retval,"\n")
}
